drivers/mspi/mspi_ambiq_ap4.c

```diff
@@ -0,0 +1,1505 @@
+	case 2048:
+		return AM_HAL_MSPI_BOUNDARY_BREAK2K;
+	case 4096:
+		return AM_HAL_MSPI_BOUNDARY_BREAK4K;
+	case 8192:
+		return AM_HAL_MSPI_BOUNDARY_BREAK8K;
+	case 16384:
+		return AM_HAL_MSPI_BOUNDARY_BREAK16K;
+	default:
+		return AM_HAL_MSPI_BOUNDARY_MAX;
+	}
+}
+
+static inline void mspi_context_ce_control(struct mspi_context *ctx, bool on)
+{
+	if (ctx->owner) {
+		if (ctx->xfer.hold_ce && ctx->xfer.ce_sw_ctrl.gpio.port != NULL) {
+			if (on) {
+				gpio_pin_set_dt(&ctx->xfer.ce_sw_ctrl.gpio, 1);
+				k_busy_wait(ctx->xfer.ce_sw_ctrl.delay);
+			} else {
+				k_busy_wait(ctx->xfer.ce_sw_ctrl.delay);
+				gpio_pin_set_dt(&ctx->xfer.ce_sw_ctrl.gpio, 0);
+			}
+		}
+	}
+}
+
+static inline void mspi_context_release(struct mspi_context *ctx)
+{
+	ctx->owner = NULL;
+	k_sem_give(&ctx->lock);
+}
+
+static inline void mspi_context_unlock_unconditionally(struct mspi_context *ctx)
+{
+	mspi_context_ce_control(ctx, false);
+
+	if (!k_sem_count_get(&ctx->lock)) {
+		ctx->owner = NULL;
+		k_sem_give(&ctx->lock);
+	}
+}
+
+static inline int mspi_context_lock(struct mspi_context *ctx, const struct mspi_dev_id *req,
+				    const struct mspi_xfer *xfer, mspi_callback_handler_t callback,
+				    struct mspi_callback_context *callback_ctx, bool lockon)
+{
+	int ret = 1;
+
+	if ((k_sem_count_get(&ctx->lock) == 0) && !lockon && (ctx->owner == req)) {
+		return 0;
+	}
+
+	if (k_sem_take(&ctx->lock, K_MSEC(xfer->timeout))) {
+		return -EBUSY;
+	}
+	if (ctx->xfer.async) {
+		if ((xfer->tx_dummy == ctx->xfer.tx_dummy) &&
+		    (xfer->rx_dummy == ctx->xfer.rx_dummy) &&
+		    (xfer->cmd_length == ctx->xfer.cmd_length) &&
+		    (xfer->addr_length == ctx->xfer.addr_length)) {
+			ret = 0;
+		} else if (ctx->packets_left == 0) {
+			if (ctx->callback_ctx) {
+				volatile struct mspi_event_data *evt_data;
+
+				evt_data = &ctx->callback_ctx->mspi_evt.evt_data;
+				while (evt_data->status != 0) {
+				}
+				ret = 1;
+			} else {
+				ret = 0;
+			}
+		} else {
+			return -EIO;
+		}
+	}
+	ctx->owner = req;
+	ctx->xfer = *xfer;
+	ctx->packets_done = 0;
+	ctx->packets_left = ctx->xfer.num_packet;
+	ctx->callback = callback;
+	ctx->callback_ctx = callback_ctx;
+	return ret;
+}
+
+static inline bool mspi_is_inp(const struct device *controller)
+{
+	struct mspi_ambiq_data *data = controller->data;
+
+	return (k_sem_count_get(&data->ctx.lock) == 0);
+}
+
+static inline int mspi_verify_device(const struct device *controller,
+				     const struct mspi_dev_id *dev_id)
+{
+	const struct mspi_ambiq_config *cfg = controller->config;
+	int device_index = cfg->mspicfg.num_periph;
+	int ret = 0;
+
+	for (int i = 0; i < cfg->mspicfg.num_periph; i++) {
+		if (dev_id->ce.port == cfg->mspicfg.ce_group[i].port &&
+		    dev_id->ce.pin == cfg->mspicfg.ce_group[i].pin &&
+		    dev_id->ce.dt_flags == cfg->mspicfg.ce_group[i].dt_flags) {
+			device_index = i;
+		}
+	}
+
+	if (device_index > cfg->mspicfg.num_periph || device_index != dev_id->dev_idx) {
```

Comment:
device_index should not equal to cfg->mspicfg.num_periph as it start from 0.
```suggestion
	if (device_index >= cfg->mspicfg.num_periph || device_index != dev_id->dev_idx) {
```